### DISCUSSION TITLE
OOM - How much ram is required for running 64 GiB sectors?

### DISCUSSION THREAD

**Discussion by jennijuju** [filecoin-project/rust-fil-proofs#1409] [Q&A] [answered]
@kernelogicL tried to pledging two 64GB sectors and it killed his 256GB RAM miner with OOM.

Another miner also mentioned that when running two PC1 at once on 3900x with 128G RAM and have `FIL_PROOFS_USE_MULTICORE_SDR=1` on, the worker was also killed by the OOM killer.

My question is, how much ram does running these proof require?

**Comment by kernelogic** [comment]
Update: once I turn off FIL_PROOFS_USE_MULTICORE_SDR, I can seal 2 64GB sectors at the same time without being OOM killed - so definitely some difference with this variable.

**Reply by jennijuju** [reply]
seems like it’s narrowed down enough for an issue ticket!

**Reply by neondragon** [reply]
I'm the miner who had the problem on 3900X, 128GB RAM, and `FIL_PROOFS_USE_MULTICORE_SDR=1`, when sealing 32GiB sectors. (@NeonixAF f019551 on Slack).
Enough swap was available, but as soon as lotus-worker hit 128GB, it was OOM killed without using the swap. I believe it's due to the multicore SDR implementation locking pages in memory, and Linux refusing to evict locked pages into swap. However, other miners told me on Slack that they were able to run two PC1 tasks in parallel with 128GB RAM and `FIL_PROOFS_USE_MULTICORE_SDR=1`. I never worked out what was different to make it work on their hardware and not on mine.

**Reply by neondragon** [reply]
I've opened an issue ticket: [https://github.com/filecoin-project/rust-fil-proofs/issues/1410](https://github.com/filecoin-project/rust-fil-proofs/issues/1410)

**Reply by cryptonemo** [reply]
@neondragon Have you dried disabling the mlock calls?  For many people they were not succeeding anyway, and in that case it maps them unlocked, which may solve your issue and properly hit swap.

**Reply by neondragon** [reply]
@cryptonemo That could very plausibly explain why some people don't encounter the issue. I've seen it on machines with nothing else running where mlock was likely to succeed.

I've discovered something else I was unaware of. I've been running the tests as root, and according to https://linux.die.net/man/2/setrlimit

> Since Linux 2.6.9, no limits are placed on the amount of memory that a privileged process may lock, and this limit instead governs the amount of memory that an unprivileged process may lock.

This may explain why my attempts to limit locked memory (ulimit -l, LimitMEMLOCK in systemd service) were unsuccessful.

I'll test with the mlock calls disabled and also as an unprivileged user and report back.

**Reply by kernelogic** [reply]
How to disable mlock calls? Where is it?

**Reply by neondragon** [reply]
@kernelogic I don't know, I grepped the code for 'mlock' and tried changing the one instance I found in storage-proofs/porep/src/stacked/vanilla/memory_handling.rs, but the process was still OOM killed.

I did find two other ways to get it to work though!

1. Run as a non-root user. 'ulimit -l' must not be unlimited.
2. Run as root using capsh to remove the memory lock capability from the process, like this
`capsh --drop=cap_ipc_lock --print -- -c "FIL_PROOFS_USE_MULTICORE_SDR=1 TMPDIR=/srv/lotus/tmp-bench  FIL_PROOFS_MAXIMIZE_CACHING=1 RUST_LOG=debug /opt/lotus/lotus-bench sealing --storage-dir /srv/lotus/seal4_1/bench --skip-commit2 --skip-unseal --num-sectors 2 --parallel 2 --sector-size 32GiB"`

Both of these methods are working for me, the process is using all the 128GB RAM and ~6.8GB swap.

I'm not completely sure if it can be called a bug in lotus or not. It seems like lotus shouldn't try to mlock more memory than the system's physical memory size. But, the mlock limit is a Linux feature. At any rate, it's really confusing!

**Reply by cryptonemo** [reply]
@kernelogic Check here: https://github.com/filecoin-project/rust-fil-proofs/blob/master/storage-proofs-porep/src/stacked/vanilla/memory_handling.rs#L262

I recommend just not running as root as @neondragon suggested.

**Comment by cryptonemo** [comment, marked as answer]
Addressed by #1413
